types/wlr_scene.c

```diff
@@ -0,0 +1,306 @@
+	}
+}
+
+struct wlr_scene *wlr_scene_create(void) {
+	struct wlr_scene *scene = calloc(1, sizeof(struct wlr_scene));
+	if (scene == NULL) {
+		return NULL;
+	}
+	scene_node_init(&scene->node, WLR_SCENE_NODE_ROOT, NULL);
+
+	return scene;
+}
+
+static void scene_surface_handle_surface_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_scene_surface *scene_surface =
+		wl_container_of(listener, scene_surface, surface_destroy);
+	wlr_scene_node_destroy(&scene_surface->node);
+}
+
+struct wlr_scene_surface *wlr_scene_surface_create(struct wlr_scene_node *parent,
+		struct wlr_surface *surface) {
+	struct wlr_scene_surface *scene_surface =
+		calloc(1, sizeof(struct wlr_scene_surface));
+	if (scene_surface == NULL) {
+		return NULL;
+	}
+	scene_node_init(&scene_surface->node, WLR_SCENE_NODE_SURFACE, parent);
+
+	scene_surface->surface = surface;
+
+	scene_surface->surface_destroy.notify = scene_surface_handle_surface_destroy;
+	wl_signal_add(&surface->events.destroy, &scene_surface->surface_destroy);
+
+	return scene_surface;
+}
+
+void wlr_scene_node_set_enabled(struct wlr_scene_node *node, bool enabled) {
+	node->state.enabled = enabled;
+}
+
+void wlr_scene_node_set_position(struct wlr_scene_node *node, int x, int y) {
+	node->state.x = x;
+	node->state.y = y;
+}
+
+void wlr_scene_node_place_above(struct wlr_scene_node *node,
+		struct wlr_scene_node *sibling) {
+	assert(node->parent == sibling->parent);
+
+	wl_list_remove(&node->state.link);
+	wl_list_insert(&sibling->state.link, &node->state.link);
+}
+
+void wlr_scene_node_place_below(struct wlr_scene_node *node,
+		struct wlr_scene_node *sibling) {
+	assert(node->parent == sibling->parent);
+
+	wl_list_remove(&node->state.link);
+	wl_list_insert(sibling->state.link.prev, &node->state.link);
+}
+
+static void scene_node_for_each_surface(struct wlr_scene_node *node,
+		int lx, int ly, wlr_surface_iterator_func_t user_iterator,
+		void *user_data) {
+	if (!node->state.enabled) {
+		return;
+	}
+
+	lx += node->state.x;
+	ly += node->state.y;
+
+	if (node->type == WLR_SCENE_NODE_SURFACE) {
+		struct wlr_scene_surface *scene_surface = scene_surface_from_node(node);
+		user_iterator(scene_surface->surface, lx, ly, user_data);
+	}
+
+	struct wlr_scene_node *child;
+	wl_list_for_each(child, &node->state.children, state.link) {
+		scene_node_for_each_surface(child, lx, ly, user_iterator, user_data);
+	}
+}
+
+void wlr_scene_node_for_each_surface(struct wlr_scene_node *node,
+		wlr_surface_iterator_func_t user_iterator, void *user_data) {
+	scene_node_for_each_surface(node, 0, 0, user_iterator, user_data);
+}
+
+struct wlr_surface *wlr_scene_node_surface_at(struct wlr_scene_node *node,
+		double lx, double ly, double *sx, double *sy) {
+	if (!node->state.enabled) {
+		return NULL;
+	}
+
+	// TODO: optimize by storing a bounding box in each node?
+	lx -= node->state.x;
+	ly -= node->state.y;
+
+	struct wlr_scene_node *child;
+	wl_list_for_each(child, &node->state.children, state.link) {
```

Comment:
Should this be `wl_list_for_each_reverse`? It looks like child nodes are stored in rendering order (bottom to top), and though the documentation comment doesn't specify, I think it's natural to assume this function would return the topmost matching surface.

Reply:
Good catch! Fixed.